Implement CEL to SQL interpreter to enhance searches

alan-ghelardi opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

<!-- Describe your changes here and link issues if any- Ideally you can get that description straight from
your descriptive commit message(s)! -->

<!-- /kind bug, cleanup, design, documentation, feature, flake, misc, question, tep -->
# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you review them:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] [Tested your changes locally](https://github.com/tektoncd/results/blob/main/docs/DEVELOPMENT.md) (if this is a code change)
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user-facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contain the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

<!--
Does your PR contain User facing changes?

If so, briefly describe them here so we can include this description in the
release notes for the next release!

For pull requests with a release note:

```release-note
Your release note here
```

For pull requests that require additional action from users switching to the new release, include the string "action required" (case insensitive) in the release note:

```release-note
action required: your release note here
```

Use the `/release-note-none` Prow command to add the `release-note-none` label to the PR for pull requests that don't need to be mentioned at release time. You can also write the string "NONE" as a release note in your PR description:

```release-note
NONE
```
-->

<!-- Feel free to add more heading to include screenshots, test logs, action items etc. Prefer removing unused options and comments to keep the description clean. -->

alan-ghelardi commented:
/kind feature

alan-ghelardi commented:
/test-all

alan-ghelardi commented:
/test all

alan-ghelardi commented:
/test all

alan-ghelardi commented:
/test all

alan-ghelardi commented:
cc @RafaeLeal

alan-ghelardi commented:
/assign @dibyom
/assign @adambkaplan

alan-ghelardi commented:
cc @tektoncd/results-maintainers

manuelwallrapp commented:
I tested this PR, it works for our requirements. Paging works, and also filtering with CEL works for us. Thank you very much for the fix @alan-ghelardi !!

alan-ghelardi commented:
/retest required

alan-ghelardi commented:
/retest-required

adambkaplan commented on pkg/api/server/v1alpha2/records.go:
```diff
@@ -63,6 +62,7 @@ func (s *Server) CreateRecord(ctx context.Context, req *pb.CreateRecordRequest)
 	// Populate Result with server provided fields.
 	protoutil.ClearOutputOnly(r)
 	r.Id = uid()
+	r.Uid = r.Id
```
good catch here. Looks like this was a lingering bug because the `.Id` field is marked "deprecated"

adambkaplan commented on pkg/api/server/v1alpha2/records.go:
```diff
@@ -142,125 +142,40 @@ func (s *Server) ListRecords(ctx context.Context, req *pb.ListRecordsRequest) (*
 	}, nil
 }
 
-// getFilteredPaginatedSortedRecords returns the specified number of results that
-// match the given CEL program.
-func (s *Server) getFilteredPaginatedSortedRecords(ctx context.Context, parent, start string, pageSize int, prg cel.Program, sortOrder string) ([]*pb.Record, error) {
```
:+1: to dropping what is likely the source of bugs my team is seeing with filtering.

adambkaplan commented on pkg/api/server/v1alpha2/records_test.go:
:+1: test updates look good!

adambkaplan commented on pkg/api/server/v1alpha2/results.go:
```diff
@@ -181,108 +179,27 @@ func (s *Server) ListResults(ctx context.Context, req *pb.ListResultsRequest) (*
 	}, nil
 }
 
-// getFilteredPaginatedSortedResults returns the specified number of results that
-// match the given CEL program.
-func (s *Server) getFilteredPaginatedSortedResults(ctx context.Context, parent string, start string, pageSize int, prg cel.Program, sortOrder string) ([]*pb.Result, error) {
```
:+1: to removing likely buggy code (and DRYing out our code base as a bonus!)

adambkaplan commented on pkg/api/server/v1alpha2/server.go:
```diff
@@ -64,12 +62,23 @@ func New(config *config.Config, logger *zap.SugaredLogger, db *gorm.DB, opts ...
 	if err != nil {
 		return nil, fmt.Errorf("failed to create CEL environment: %w", err)
 	}
+	// TODO: turn the func into a MustX that should panic on error.
+	resultsEnv, err := resultscel.NewResultsEnv()
```
Noting in this comment for future reference (as a GH issue)

adambkaplan commented on test/e2e/e2e_test.go:
```diff
@@ -505,15 +505,15 @@ func TestListResults(t *testing.T) {
 		want, err := gc.ListResults(ctx, &resultsv1alpha2.ListResultsRequest{
 			Parent:  "default",
-			OrderBy: "created_time",
+			OrderBy: "create_time",
```
Do we need to note this as a "breaking" change? I'm fine as long as we have an appropriate "ACTION REQUIRED" release note.

alan-ghelardi replied:
Yep, absolutely. Since create_time and update_ttime are deprecated in the proto schema, I decided to change the fields in the order  by clause as well to simplify the API from the user perspective. Then, people do not need to perform a mental mapping between the wire representation and thee database model; they may use the same wire fields to query the API. I'll add to the release notes.

adambkaplan reviewed (approved):
/approve

I feel like I owe an apology for not reviewing this until now. This is a phenomenally well crafted PR and will likely address critical bugs my team is observing with the CEL filtering. Well done!

I have two requests before we merge this code:

1. Update the commit message body with the contents of the paragraph in the PR description below the release note (or a shortened version thereof). This is a very significant code change, and commit messages help future contributors understand the deeper "why" behind the updates. See [commit message standards](https://github.com/tektoncd/community/blob/main/standards.md#commit-messages) for further guidance.
2. Add an "ACTION REQUIRED" release note if the filtering by create timestamp (or other field) has changed.

Everything else (docs, nits, etc.) we should address in follow up issues or pull requests.

alan-ghelardi commented on pkg/api/server/v1alpha2/lister/proto/page_token.proto:
I kept there to follow the same pattern as the proto files at pkg/api/server/db/pagination/proto , but agree that makes sense grouping them in the proto folder at the root directory. I'll refactor in a follow-up pr then.

alan-ghelardi commented:
@adambkaplan  I believe that I managed to address all of your requesstss.

@dibyom I'd love to see your feedback as well to move forward with this featture.

adambkaplan commented:
/retest

avinal reviewed (approved):
Here are the results of manual tests for filtering and ordering I did, with 100% accurate results.

avinal commented:
Awesome work @alan-ghelardi 🎆

dibyom commented:
@alan-ghelardi I did not get a chance to go in depth with everything in this PR but from going though the cel2sql package and at a high level through the rest of the changes, this looks great! Really nice work (and the PR is very readable for the size as well!)
Given that there are no major API changes, I +1 for merging this!

/approve

adambkaplan commented:
/lgtm

:tada: :tada: :tada:

alan-ghelardi commented:
Well, looks like that the GH action job is blocking the merge. I saw that there's an ongoing attempt to fix the issue at https://github.com/tektoncd/results/pull/492. Any suggestion for getting around this? Maybe we could delete the Actions file in this branch while the issue isn't fixed?

alan-ghelardi commented:
@khrm  and @enarha I think that you all were on that ☝️

enarha commented:
It's not clear to me if the CodeQL is actually required/blocking, it's not marked as such, but maybe I'm missing something. I thought the check was also disabled which moved the task of fixing it lower in my TODO list, but I'll be back on that today.

enarha commented:
/retest

alan-ghelardi commented:
I've rebased onto main and deleted the GitHub Action workflow temporarily to avoid blocking the merge for now. We can restore it once https://github.com/tektoncd/results/pull/492 is done. Wdyt?  A new lgtm will be required here. cc @adambkaplan .

avinal commented:
IMO that should be fine.

avinal reviewed (approved):
/lgtm

dibyom commented:
/lgtm
